numpy/lib/polynomial.py

```diff
@@ -678,7 +678,10 @@ def polyval(p, x):
         x = NX.asarray(x)
         y = NX.zeros_like(x)
     for i in range(len(p)):
-        y = y * x + p[i]
+        # separate out the Horner operation y = y*x + p[i],
+        # so there are fewer memory allocations
+        y *= x
```

Comment:
What happens here if `x` is integer and `p` is float? I think one needs to ensure y gets set up right...